hal/src/nRF52840/i2c_hal.cpp

```diff
@@ -636,29 +648,57 @@ uint8_t hal_i2c_reset(hal_i2c_interface_t i2c, uint32_t reserved, void* reserved
     }
 
     I2cLock lk(i2c);
-    if (hal_i2c_is_enabled(i2c, nullptr)) {
-        hal_i2c_end(i2c, nullptr);
+    if (!hal_i2c_is_enabled(i2c, nullptr) || (i2cMap[i2c].mode != I2C_MODE_MASTER)) {
+        return 1;
+    }
+
+    // Important: we keep GPIO configuration intact
+    if (twiUninit(i2c, false) == SYSTEM_ERROR_NONE) {
+        i2cMap[i2c].state = HAL_I2C_STATE_DISABLED;
+    }
+
+    // Just in case make sure that the pins are correctly configured (they should anyway be at this point)
+    hal_gpio_config_t conf = {
+        .size = sizeof(conf),
+        .version = HAL_GPIO_VERSION,
+        .mode = OUTPUT_OPEN_DRAIN_PULLUP,
+        .set_value = true,
+        .value = 1
+    };
+    HAL_Pin_Configure(i2cMap[i2c].sda_pin, &conf, nullptr);
+    conf.value = HAL_GPIO_Read(i2cMap[i2c].scl_pin);
+    HAL_Pin_Configure(i2cMap[i2c].scl_pin, &conf, nullptr);
 
-        HAL_Pin_Mode(i2cMap[i2c].sda_pin, INPUT_PULLUP); //Turn SCA into high impedance input
-        HAL_Pin_Mode(i2cMap[i2c].scl_pin, OUTPUT);       //Turn SCL into a normal GPO
-        HAL_GPIO_Write(i2cMap[i2c].scl_pin, 1);     // Start idle HIGH
+    // Generate up to 9 pulses on SCL to tell slave to release the bus
+    for (int i = 0; i < 9; i++) {
+        HAL_GPIO_Write(i2cMap[i2c].sda_pin, 1);
+        HAL_Delay_Microseconds(50);
 
-        //Generate 9 pulses on SCL to tell slave to release the bus
-        for (int i = 0; i < 9; i++) {
+        if (HAL_GPIO_Read(i2cMap[i2c].sda_pin) == 0) {
             HAL_GPIO_Write(i2cMap[i2c].scl_pin, 0);
-            HAL_Delay_Microseconds(100);
+            HAL_Delay_Microseconds(50);
             HAL_GPIO_Write(i2cMap[i2c].scl_pin, 1);
-            HAL_Delay_Microseconds(100);
+            HAL_Delay_Microseconds(50);
+            HAL_GPIO_Write(i2cMap[i2c].scl_pin, 0);
```

Comment:
Does it matter if this is left low as the last transition?

Reply:
Doesn't, because generating a stop condition below will anyway pull it correctly high.